### DISCUSSION TITLE
Disable Button while Data Source is loading. Any help ?

### DISCUSSION THREAD

**Discussion by poirazis** [Budibase/budibase#7099] [Help] [answered]
Is there a hackish way I can disable a button until one or more data sources are finished loading ?
Like a binding {{ Data Source.IsLoading }} kind of thing.

**Comment by R2bEEaton** [comment, marked as answer]
Hello!

### Easy
I think the easiest way to disable a button until data sources are finished loading is to include the button inside of the data source(s) in the component tree, like so:
![image](https://user-images.githubusercontent.com/34921506/183395326-0d7cd2d1-073b-4c3f-8b3e-83366152fafd.png)

### "Hackish"
Anything inside the data provider element will not be added to the DOM until the source finishes loading. Now, if you want to disable a button outside of it, here's a "hackish" way to do so:

1. Create a structure like this.
![image](https://user-images.githubusercontent.com/34921506/183399549-44840029-ba59-4168-8347-2c4a177a884c.png)

2. Inside the embed, put the following where you replace "button-to-enable" with whatever you set the "Name" attribute of the button to that you want to enable when the data source loads.
```
<img style="display: none;" src onerror="
	document.querySelectorAll('div[data-name=&quot;button-to-enable&quot;]')[0].children[0].disabled = false;
">
```

Elements inside Data Sources only load after the source itself finishes loading. Therefore, the embed will get loaded at this point and then immediately trigger a JavaScript function using XSS (#6839) which will un-disable the button we specify.

Here's an example export if you want to examine it further.
[Test-export-1659955281137.txt](https://github.com/Budibase/budibase/files/9281070/Test-export-1659955281137.txt)

This works for as many data sources as you want, as long as they are all nested and the embed element is inside the last one.
```
Source
    -> Source
        -> Source
            -> Embed
```


### "Hackish" for multiple un-nested data sources
Similarly to the single data source method, you could come up with a variable to count the number of data sources loaded so far and then only enable the button when that count reaches the desired number.

Include the following embed element in each un-nested data source you want to use to enable the button, changing the desired `loadedcount` to the total count of these data sources:
```
<img style="display: none;" src onerror="
	if (typeof loadedcount == 'undefined') {
		loadedcount = 1;
	} else {
		loadedcount++;
		if (loadedcount == 2) {
			document.querySelectorAll('div[data-name=&quot;button-to-enable&quot;]')[0].children[0].disabled = false;
		}
	}
">
```
![image](https://user-images.githubusercontent.com/34921506/183402114-f958f419-cc04-4b76-9849-23b4a9a7498c.png)

And here's an export of that.
[Test-export-1659956044652.txt](https://github.com/Budibase/budibase/files/9281145/Test-export-1659956044652.txt)

### Good luck!

**Reply by poirazis** [reply]
Thank you so much @R2bEEaton !
